msprime/mutations.py

```diff
@@ -199,6 +199,308 @@ def __init__(self, state_independent=False):
+                )
+            elif i == j + 1:
+                rate_matrix[ii, jj] = (
+                    beta_microsat(i, s, lo)
+                    * (1 - alpha_microsat(i, u, v, lo))
+                    * (p + ((1 - p) * gamma_microsat(m, i, j, lo, hi)))
+                )
+            elif i > j + 1:
+                rate_matrix[ii, jj] = (
+                    beta_microsat(i, s, lo)
+                    * (1 - alpha_microsat(i, u, v, lo))
+                    * (1 - p)
+                    * gamma_microsat(m, i, j, lo, hi)
+                )
+            else:
+                rate_matrix[ii, jj] = 0
+    # scale by max row sum; then normalize to 1
+    row_sums = rate_matrix.sum(axis=1, dtype="float64")
+    alpha = max(row_sums)
+    rate_matrix /= alpha
+    row_sums = rate_matrix.sum(axis=1, dtype="float64")
+    np.fill_diagonal(rate_matrix, 1.0 - row_sums)
+    return rate_matrix
+
+
+class MicrosatMutationModel(MatrixMutationModel):
+    """
+    General class for Microsatellite mutation models
+    which parameterizes the transition matrix describing
+    changes among alleles using 5 parameters according
+    to the model of Sainudiin et al. (2004).
+
+    Concretely, mutation rates under this model are computed as follows.
+    Let ``qij`` be the transition rate from equation (3) in Sainudiin et al,
+    and let ``M`` be the largest row sum of the matrix whose i,j-th entry
+    is ``qij`` and whose diagonals are zero. Then if the mutation rate
+    passed to :func:`.sim_mutations` is ``m``, the mutation rate from
+    allele ``i`` to allele `j`` is ``m * qij / M``.
+
+    :param float s: strength of length dependence on mutation rate
+        (i.e. s=0 uniform rate)
+    :param float u: constant bias parameter; must be between 0 and 1
+    :param float v: linear bias parameter
+    :param float p: probability of a single step mutation
+    :param float m: 1 / (mean multistep mutation size)
+    :param int lo: lower bound on repeat number
+    :param int hi: upper bound on repeat number
+    :param list[float] root_distribution: An array of float values of length hi-lo+1
+        which should sum to 1, and give the probabilities used to choose
+        the ancestral state of each microsatellite.
+        Defaults to the stationary distribution of the model.
+
+    """
+
+    def __init__(
+        self,
+        s=0,
+        u=0.5,
+        v=0.0,
+        p=1.0,
+        m=1.0,
+        lo=2,
+        hi=50,
+        root_distribution=None,
+    ):
+        if np.isinf(lo):
+            raise ValueError("lo is infinite")
+        if np.isinf(hi):
+            raise ValueError("hi is infinite")
+        if lo > hi:
+            raise ValueError("lo > hi")
+        if not lo >= 0:
+            raise ValueError("lo must be >= 0")
+        if s <= -1 / (hi - lo + 1):
+            raise ValueError("s must be > -1/(hi-lo+1)")
+        if u < 0 or u > 1:
+            raise ValueError("u must be between 0 and 1 inclusive")
```

Comment:
I think this should be *non*-inclusive. Otherwise, some wierd things can happen, e.g. the chain is not ergodic:
```
>>> mm = msprime.MicrosatMutationModel(u=0, lo=3, hi=6, v=-10)
>>> mm.transition_matrix
array([[1., 0., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 1.]])
>>> mm.root_distribution
array([1., 0., 0., 0.])

```

Reply:
we can change this but i think it conflicts with the original paper
<img width="483" alt="image" src="https://user-images.githubusercontent.com/7674197/175462652-dd9262bc-2881-440d-a123-215c9435a629.png">

Reply:
Well, you get a valid Markov chain for the endpoints, but where one or both of the endpoints are absorbing, and so you'd get a situation where the root distribution draws with probability 1 from an absorbing state, and so you'd never get any non-silent mutations. That seems like not what anyone would actually want? And, the derived models have 0 < u < 1? There's no harm in leaving it 0 <= u <= 1, I guess, but I'd vote to restrict it.